[baseline-alignment][flex] Ascent for flex items that are scroll containers should be clamped to the border box.

sammygill opened the pull request:
#### 1e66b0fc954597d3d357d1c7db0c7b8373dad8b1
<pre>
[baseline-alignment][flex] Ascent for flex items that are scroll containers should be clamped to the border box.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246229">https://bugs.webkit.org/show_bug.cgi?id=246229</a>
rdar://100908615

Reviewed by Alan Baradlay.

If a flex item is a scroll container (e.g. overflow scroll/hidden) then
it may provide an ascent value that is outside the visible content box
and result in a non-visible position being used for baseline alignment.

When determining the ascent of a flex item, we will check to see if it
is a scroll container by checking to see if it has an overflow value
of either scroll, hidden, or auto and then clamp the used ascent value
of the box to the bottom/top edges of its border box if it is.

<a href="https://drafts.csswg.org/css-align/#baseline-export">https://drafts.csswg.org/css-align/#baseline-export</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-line-clamp-001.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-overflow-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-overflow-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-overflow-003-expected.txt:
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::isScrollContainerX const):
(WebCore::RenderBox::isScrollContainerY const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::flexItemIsScrollContainerCrossAxis const):
(WebCore::RenderFlexibleBox::marginBoxAscentForChild):
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren):
* Source/WebCore/rendering/RenderFlexibleBox.h:

Canonical link: <a href="https://commits.webkit.org/263974@main">https://commits.webkit.org/263974@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5a5833047c5e2cd0ec69950f368cfdbb53d368e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7730 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6504 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6167 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6575 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6307 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9383 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6277 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6295 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5582 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7797 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3769 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5562 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13458 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5634 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5641 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7878 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6112 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5000 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5529 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1486 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9673 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5896 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1013,6 +1013,11 @@ LayoutUnit RenderFlexibleBox::crossAxisScrollbarExtent() const
     return isHorizontalFlow() ? horizontalScrollbarHeight() : verticalScrollbarWidth();
 }
 
+bool RenderFlexibleBox::childHasScrollableOverflowCrossAxis(const RenderBox& child) const
```
This name is a bit misleading as just because the flex item is a scroll container, it does not necessarily mean it has scrollable overflow (it may or may not have). -it should also read flexItemHasFoobar since what we have here is (always) a flex item.

sammygill replied:
Ah, right changed to `flexItemIsScrollContainerCrossAxis` The flexItem portion of the name does make it inconsistent with the other helpers in the class FWIW

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1493,9 +1498,15 @@ LayoutUnit RenderFlexibleBox::marginBoxAscentForChild(const RenderBox& child)
+    auto childMarginBefore = flowAwareMarginBeforeForChild(child);
+    if (!ascent) {
+        if (childHasScrollableOverflowCrossAxis(child))
+            return std::clamp(synthesizedBaseline(child, style(), direction, BorderBox) + childMarginBefore, childMarginBefore, childMarginBefore + crossAxisExtentForChild(child));
```
isn't this just
`std::clamp(synthesizedBaseline(child, style(), direction, BorderBox), 0, crossAxisExtentForChild(child)) + childMarginBefore;`
and then the question is whether synthesizedBaseline() could really be < 0 as if not, it's just a `std::min(synthesizedBaseline(child, style(), direction, BorderBox), crossAxisExtentForChild(child)) + childMarginBefore;`

sammygill replied:
Yes it is, but I actually don't think we need to do anything special here now that I look at it again. Since the synthesized baseline is based off the border box of the flex item, I think it should always provide a value inside the visible portion of the box even it is a scroll container.

For the similar scenario below, I think we still need to keep the logic (and I applied the change there also), since its based off the content of the flex item

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1013,6 +1013,11 @@ LayoutUnit RenderFlexibleBox::crossAxisScrollbarExtent() const
+{
+    return isHorizontalFlow() ? child.isScrollContainerY() : child.isScrollContainerX();
+}
+
```
Does this need a dedicated function?

sammygill replied:
Not necessarily. Was mostly just for readability but after inlining the logic into the call site its still seems pretty readable

sammygill committed:
[baseline-alignment][flex] Ascent for flex items that are scroll containers should be clamped to the border box.
https://bugs.webkit.org/show_bug.cgi?id=246229
rdar://100908615

Reviewed by Alan Baradlay.

If a flex item is a scroll container (e.g. overflow scroll/hidden) then
it may provide an ascent value that is outside the visible content box
and result in a non-visible position being used for baseline alignment.

When determining the ascent of a flex item, we will check to see if it
is a scroll container by checking to see if it has an overflow value
of either scroll, hidden, or auto and then clamp the used ascent value
of the box to the bottom/top edges of its border box if it is.

https://drafts.csswg.org/css-align/#baseline-export

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-line-clamp-001.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-overflow-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-overflow-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-overflow-003-expected.txt:
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::isScrollContainerX const):
(WebCore::RenderBox::isScrollContainerY const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::flexItemIsScrollContainerCrossAxis const):
(WebCore::RenderFlexibleBox::marginBoxAscentForChild):
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren):
* Source/WebCore/rendering/RenderFlexibleBox.h:

Canonical link: https://commits.webkit.org/263974@main
